Avoid returning null in GTFSFeed

Instead of just logging we should throw an exception or would this break some code?

https://github.com/graphhopper/graphhopper/blob/master/reader-gtfs/src/main/java/com/conveyal/gtfs/GTFSFeed.java#L368

Related [discussion](https://discuss.graphhopper.com/t/issues-setting-up-gtfs-routing/3978), the full stacktrace is:

```
ERROR [2019-01-10 08:54:19,413] com.conveyal.gtfs.GTFSFeed: Could not construct db from file.
! java.io.IOException: Wrong index checksum, store was not closed properly and could be corrupted.
! ... 18 common frames omitted
! Causing: java.io.IOError: java.io.IOException: Wrong index checksum, store was not closed properly and could be corrupted.
! at org.mapdb.StoreDirect.checkHeaders(StoreDirect.java:269)
! at org.mapdb.StoreDirect.<init>(StoreDirect.java:207)
! at org.mapdb.DBMaker.extendStoreDirect(DBMaker.java:971)
! at org.mapdb.DBMaker.makeEngine(DBMaker.java:758)
! at org.mapdb.DBMaker.make(DBMaker.java:701)
! at com.conveyal.gtfs.GTFSFeed.constructDB(GTFSFeed.java:365)
! at com.conveyal.gtfs.GTFSFeed.<init>(GTFSFeed.java:352)
! at com.graphhopper.reader.gtfs.GtfsStorage.loadGtfsFromFile(GtfsStorage.java:198)
! at com.graphhopper.reader.gtfs.GraphHopperGtfs.createOrLoad(GraphHopperGtfs.java:347)
! at com.graphhopper.http.GraphHopperBundle.runPtGraphHopper(GraphHopperBundle.java:213)
! at com.graphhopper.http.GraphHopperBundle.run(GraphHopperBundle.java:202)
! at com.graphhopper.http.GraphHopperBundle.run(GraphHopperBundle.java:64)
! at io.dropwizard.setup.Bootstrap.run(Bootstrap.java:200)
! at io.dropwizard.cli.EnvironmentCommand.run(EnvironmentCommand.java:42)
! at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:87)
! at io.dropwizard.cli.Cli.run(Cli.java:78)
! at io.dropwizard.Application.run(Application.java:93)
! at com.graphhopper.http.GraphHopperApplication.main(GraphHopperApplication.java:33)
java.lang.RuntimeException: java.lang.NullPointerException
	at com.graphhopper.reader.gtfs.GtfsStorage.loadGtfsFromFile(GtfsStorage.java:203)
	at com.graphhopper.reader.gtfs.GraphHopperGtfs.createOrLoad(GraphHopperGtfs.java:347)
	at com.graphhopper.http.GraphHopperBundle.runPtGraphHopper(GraphHopperBundle.java:213)
	at com.graphhopper.http.GraphHopperBundle.run(GraphHopperBundle.java:202)
	at com.graphhopper.http.GraphHopperBundle.run(GraphHopperBundle.java:64)
	at io.dropwizard.setup.Bootstrap.run(Bootstrap.java:200)
	at io.dropwizard.cli.EnvironmentCommand.run(EnvironmentCommand.java:42)
	at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:87)
	at io.dropwizard.cli.Cli.run(Cli.java:78)
	at io.dropwizard.Application.run(Application.java:93)
	at com.graphhopper.http.GraphHopperApplication.main(GraphHopperApplication.java:33)
Caused by: java.lang.NullPointerException
	at com.conveyal.gtfs.GTFSFeed.<init>(GTFSFeed.java:376)
	at com.conveyal.gtfs.GTFSFeed.<init>(GTFSFeed.java:352)
	at com.graphhopper.reader.gtfs.GtfsStorage.loadGtfsFromFile(GtfsStorage.java:198)
	... 10 more
```

--- a/reader-gtfs/src/main/java/com/conveyal/gtfs/GTFSFeed.java
+++ b/reader-gtfs/src/main/java/com/conveyal/gtfs/GTFSFeed.java
@@ -27,10 +27,9 @@
 package com.conveyal.gtfs;
 
 import com.conveyal.gtfs.error.GTFSError;
-import com.conveyal.gtfs.model.*;
 import com.conveyal.gtfs.model.Calendar;
+import com.conveyal.gtfs.model.*;
 import com.google.common.collect.Iterables;
-import com.google.common.util.concurrent.ExecutionError;
 import org.locationtech.jts.geom.Coordinate;
 import org.locationtech.jts.geom.CoordinateList;
 import org.locationtech.jts.geom.GeometryFactory;
@@ -45,13 +44,11 @@ import org.slf4j.LoggerFactory;
 
 import java.io.Closeable;
 import java.io.File;
-import java.io.IOError;
 import java.io.IOException;
 import java.time.LocalDate;
 import java.time.format.DateTimeFormatter;
 import java.util.*;
 import java.util.concurrent.ConcurrentNavigableMap;
-import java.util.concurrent.ExecutionException;
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
 import java.util.zip.ZipEntry;
@@ -111,7 +108,7 @@ public class GTFSFeed implements Cloneable, Closeable {
      *
      * Interestingly, all references are resolvable when tables are loaded in alphabetical order.
      */
-    public void loadFromFile(ZipFile zip, String fid) throws Exception {
+    public void loadFromFile(ZipFile zip, String fid) throws IOException {
         if (this.loaded) throw new UnsupportedOperationException("Attempt to load GTFS into existing database");
 
         // NB we don't have a single CRC for the file, so we combine all the CRCs of the component files. NB we are not
@@ -170,7 +167,7 @@ public class GTFSFeed implements Cloneable, Closeable {
         loaded = true;
     }
 
-    public void loadFromFile(ZipFile zip) throws Exception {
+    public void loadFromFile(ZipFile zip) throws IOException {
         loadFromFile(zip, null);
     }
 
@@ -348,26 +345,17 @@ public class GTFSFeed implements Cloneable, Closeable {
     }
 
     /** Create a GTFS feed connected to a particular DB, which will be created if it does not exist. */
-    public GTFSFeed (String dbFile) throws IOException, ExecutionException {
-        this(constructDB(dbFile)); // TODO db.close();
+    public GTFSFeed(File file) {
+        this(constructDB(file));
     }
 
-    private static DB constructDB(String dbFile) {
-        DB db;
-        try{
-            DBMaker dbMaker = DBMaker.newFileDB(new File(dbFile));
-            db = dbMaker
-                    .transactionDisable()
-                    .mmapFileEnable()
-                    .asyncWriteEnable()
-                    .compressionEnable()
-//                     .cacheSize(1024 * 1024) this bloats memory consumption
-                    .make();
-            return db;
-        } catch (ExecutionError | IOError | Exception e) {
-            LOG.error("Could not construct db from file.", e);
-            return null;
-        }
+    private static DB constructDB(File file) {
+        return DBMaker.newFileDB(file)
+                .transactionDisable()
+                .mmapFileEnable()
+                .asyncWriteEnable()
+                .compressionEnable()
+                .make();
     }
 
     private GTFSFeed (DB db) {

--- a/reader-gtfs/src/main/java/com/conveyal/gtfs/GTFSFeed.java
+++ b/reader-gtfs/src/main/java/com/conveyal/gtfs/GTFSFeed.java
@@ -171,8 +171,11 @@ public class GTFSFeed implements Cloneable, Closeable {
         loaded = true;
     }
 
-    public void loadFromFile(ZipFile zip) throws IOException {
+    public void loadFromFileAndLogErrors(ZipFile zip) throws IOException {
         loadFromFile(zip, null);
+        for (GTFSError error : errors) {
+            LOG.error(error.getMessageWithContext());
+        }
     }
 
     public boolean hasFeedInfo () {

--- a/reader-gtfs/src/main/java/com/conveyal/gtfs/GTFSFeed.java
+++ b/reader-gtfs/src/main/java/com/conveyal/gtfs/GTFSFeed.java
@@ -27,6 +27,7 @@
 package com.conveyal.gtfs;
 
 import com.conveyal.gtfs.error.GTFSError;
+import com.conveyal.gtfs.error.GeneralError;
 import com.conveyal.gtfs.model.Calendar;
 import com.conveyal.gtfs.model.*;
 import com.google.common.collect.Iterables;
@@ -140,6 +141,9 @@ public class GTFSFeed implements Cloneable, Closeable {
         db.getAtomicString("feed_id").set(feedId);
 
         new Agency.Loader(this).loadTable(zip);
+        if (agency.isEmpty()) {
+            errors.add(new GeneralError("agency", 0, "agency_id", "Need at least one agency."));
+        }
 
         // calendars and calendar dates are joined into services. This means a lot of manipulating service objects as
         // they are loaded; since mapdb keys/values are immutable, load them in memory then copy them to MapDB once

--- a/reader-gtfs/src/main/java/com/conveyal/gtfs/error/GeneralError.java
+++ b/reader-gtfs/src/main/java/com/conveyal/gtfs/error/GeneralError.java
@@ -36,6 +36,7 @@ public class GeneralError extends GTFSError implements Serializable {
 
     public GeneralError(String file, long line, String field, String message) {
         super(file, line, field);
+        this.message = message;
     }
 
     @Override public String getMessage() {

--- a/reader-gtfs/src/main/java/com/conveyal/gtfs/model/Route.java
+++ b/reader-gtfs/src/main/java/com/conveyal/gtfs/model/Route.java
@@ -76,11 +76,14 @@ public class Route extends Entity { // implements Entity.Factory<Route>
             Agency agency = getRefField("agency_id", false, feed.agency);
 
             // if there is only one agency, associate with it automatically
-            // TODO: what will this do if the agency and the route have agency_ids but they do not match?
-            if (agency == null && feed.agency.size() == 1)
+            if (agency == null && feed.agency.size() == 1) {
                 agency = feed.agency.values().iterator().next();
+            }
 
-            r.agency_id = agency.agency_id;
+            // If we still haven't got an agency, it's because we have a bad reference, or because there is no agency
+            if (agency != null) {
+                r.agency_id = agency.agency_id;
+            }
 
             r.route_short_name = getStringField("route_short_name", false); // one or the other required, needs a special validator
             r.route_long_name = getStringField("route_long_name", false);

--- a/reader-gtfs/src/main/java/com/graphhopper/reader/gtfs/GraphHopperGtfs.java
+++ b/reader-gtfs/src/main/java/com/graphhopper/reader/gtfs/GraphHopperGtfs.java
@@ -366,7 +366,11 @@ public final class GraphHopperGtfs implements GraphHopperAPI {
                             t.transfer.min_transfer_time = (int) (t.time / 1000L);
                             gtfsFeed.transfers.put(t.id, t.transfer);
                         });
-                gtfsReader.buildPtNetwork();
+                try {
+                    gtfsReader.buildPtNetwork();
+                } catch (Exception e) {
+                    throw new RuntimeException("Error while constructing transit network. Is your GTFS file valid? Please check log for possible causes.", e);
+                }
             }
             graphHopperStorage.flush();
             return graphHopperStorage;

--- a/reader-gtfs/src/main/java/com/graphhopper/reader/gtfs/GtfsStorage.java
+++ b/reader-gtfs/src/main/java/com/graphhopper/reader/gtfs/GtfsStorage.java
@@ -26,14 +26,18 @@ import com.graphhopper.gtfs.fare.FixedFareAttributeLoader;
 import com.graphhopper.storage.Directory;
 import com.graphhopper.storage.Graph;
 import com.graphhopper.storage.GraphExtension;
-import org.mapdb.*;
+import org.mapdb.Bind;
+import org.mapdb.DB;
+import org.mapdb.DBMaker;
+import org.mapdb.HTreeMap;
 
-import java.io.*;
+import java.io.File;
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.file.Files;
 import java.time.LocalDate;
 import java.time.ZoneId;
-import java.nio.file.Files;
 import java.util.*;
-import java.util.concurrent.ExecutionException;
 import java.util.zip.ZipFile;
 
 public class GtfsStorage implements GraphExtension, GtfsStorageI {
@@ -150,13 +154,9 @@ public class GtfsStorage implements GraphExtension, GtfsStorageI {
 		this.data = DBMaker.newFileDB(new File(dir.getLocation() + "/transit_schedule")).transactionDisable().mmapFileEnable().readOnly().make();
 		init();
 		for (String gtfsFeedId : this.gtfsFeedIds) {
-			try {
-				GTFSFeed feed = new GTFSFeed(dir.getLocation() + "/" + gtfsFeedId);
-				this.gtfsFeeds.put(gtfsFeedId, feed);
-				this.transfers.put(gtfsFeedId, new Transfers(feed));
-			} catch (IOException | ExecutionException e) {
-				throw new RuntimeException(e);
-			}
+			GTFSFeed feed = new GTFSFeed(new File(dir.getLocation() + "/" + gtfsFeedId));
+			this.gtfsFeeds.put(gtfsFeedId, feed);
+			this.transfers.put(gtfsFeedId, new Transfers(feed));
 		}
 		return true;
 	}
@@ -194,12 +194,14 @@ public class GtfsStorage implements GraphExtension, GtfsStorageI {
 	}
 
 	void loadGtfsFromFile(String id, ZipFile zip) {
+		File file = new File(dir.getLocation() + "/" + id);
 		try {
-			GTFSFeed feed = new GTFSFeed(dir.getLocation() + "/" + id);
+			Files.deleteIfExists(file.toPath());
+			GTFSFeed feed = new GTFSFeed(file);
 			feed.loadFromFile(zip);
 			fixFares(feed, zip);
 			this.gtfsFeeds.put(id, feed);
-		} catch (Exception e) {
+		} catch (IOException e) {
 			throw new RuntimeException(e);
 		}
 		this.gtfsFeedIds.add(id);

--- a/reader-gtfs/src/main/java/com/graphhopper/reader/gtfs/GtfsStorage.java
+++ b/reader-gtfs/src/main/java/com/graphhopper/reader/gtfs/GtfsStorage.java
@@ -19,7 +19,6 @@
 package com.graphhopper.reader.gtfs;
 
 import com.conveyal.gtfs.GTFSFeed;
-import com.conveyal.gtfs.error.GTFSError;
 import com.conveyal.gtfs.model.Fare;
 import com.conveyal.gtfs.model.FareRule;
 import com.google.transit.realtime.GtfsRealtime;
@@ -199,10 +198,7 @@ public class GtfsStorage implements GraphExtension, GtfsStorageI {
 		try {
 			Files.deleteIfExists(file.toPath());
 			GTFSFeed feed = new GTFSFeed(file);
-			feed.loadFromFile(zip);
-			for (GTFSError error : feed.errors) {
-				System.out.println(error.getMessageWithContext());
-			}
+			feed.loadFromFileAndLogErrors(zip);
 			fixFares(feed, zip);
 			this.gtfsFeeds.put(id, feed);
 		} catch (IOException e) {

--- a/reader-gtfs/src/main/java/com/graphhopper/reader/gtfs/GtfsStorage.java
+++ b/reader-gtfs/src/main/java/com/graphhopper/reader/gtfs/GtfsStorage.java
@@ -19,6 +19,7 @@
 package com.graphhopper.reader.gtfs;
 
 import com.conveyal.gtfs.GTFSFeed;
+import com.conveyal.gtfs.error.GTFSError;
 import com.conveyal.gtfs.model.Fare;
 import com.conveyal.gtfs.model.FareRule;
 import com.google.transit.realtime.GtfsRealtime;
@@ -199,6 +200,9 @@ public class GtfsStorage implements GraphExtension, GtfsStorageI {
 			Files.deleteIfExists(file.toPath());
 			GTFSFeed feed = new GTFSFeed(file);
 			feed.loadFromFile(zip);
+			for (GTFSError error : feed.errors) {
+				System.out.println(error.getMessageWithContext());
+			}
 			fixFares(feed, zip);
 			this.gtfsFeeds.put(id, feed);
 		} catch (IOException e) {

